extensions/standard-processors/processors/InvokeHTTP.h

```diff
@@ -41,9 +43,59 @@
 #include "utils/Enum.h"
 #include "utils/RegexUtils.h"
 
+namespace org::apache::nifi::minifi::test {
+class HttpClientStoreTestAccessor;
+}
+
 namespace org::apache::nifi::minifi::processors {
 
 namespace invoke_http {
+class HttpClientStore {
+ public:
+  HttpClientStore(const size_t size, std::function<gsl::not_null<std::unique_ptr<minifi::http::HTTPClient>>(const std::string&)> create_client_function)
+      : max_size_(size),
+        create_client_function_(std::move(create_client_function)) {
+  }
+  HttpClientStore(const HttpClientStore&) = delete;
+  HttpClientStore& operator=(const HttpClientStore&) = delete;
+  HttpClientStore(HttpClientStore&&) = delete;
+  HttpClientStore& operator=(HttpClientStore&&) = delete;
+  ~HttpClientStore() = default;
+
+  class HttpClientWrapper {
+   public:
+    HttpClientWrapper(HttpClientStore& client_store, minifi::http::HTTPClient& client) : client_(client), client_store_(client_store) {
+    }
+    HttpClientWrapper(HttpClientWrapper&& src) = default;
+    HttpClientWrapper(const HttpClientWrapper&) = delete;
+    ~HttpClientWrapper() {
+      client_store_.returnClient(client_);
+    }
+
+    minifi::http::HTTPClient& get() const {
+      return client_;
+    }
+
+   private:
+    minifi::http::HTTPClient& client_;
+    HttpClientStore& client_store_;
+  };
+
+  [[nodiscard]] HttpClientWrapper getClient(const std::string& url);
+
+ private:
+  friend class ::org::apache::nifi::minifi::test::HttpClientStoreTestAccessor;
+  void returnClient(minifi::http::HTTPClient& client);
+
+  std::mutex clients_mutex_;
+  std::condition_variable cv_;
+  const size_t max_size_;
+  std::list<gsl::not_null<std::unique_ptr<minifi::http::HTTPClient>>> used_clients_;
+  std::list<gsl::not_null<std::unique_ptr<minifi::http::HTTPClient>>> unused_clients_;
+  std::function<gsl::not_null<std::unique_ptr<minifi::http::HTTPClient>>(const std::string&)> create_client_function_;
```

Comment:
I think we could do without the extra indirection of unique_ptr. List splicing doesn't move the data on the heap.

Reply:
That's true, unfortunately we would still need to copy or move the HTTPClient on line 56 and line 176 and those constructors are delete from the HTTPClient class.

Reply:
line 56: It could be an in-place constructed to the list. Emplace and returning it by value from the create function should be able to do that.
line 176: return by value

Reply:
I checked and returning by value doesn't seem to work without copy/move. I thought it could be elided. So nevermind, leave them as pointers.